### DISCUSSION TITLE
Build a complex site with several parts in common. Templates?

### DISCUSSION THREAD

**Discussion by JohnRobson** [fastify/fastify#2585] [General]
This is a newbie question. I'm a Flask developer trying to migrate to Node.js/Fastify.

To build a complex site with several parts in common, in Flask we can use "Template Inheritance" [1] and "Blueprints" [2].

What is the correct approach in Node.js/Fastify?  Is there some tutorial about this?

Thank you.

[1] https://flask.palletsprojects.com/en/1.1.x/patterns/templateinheritance/

[2] https://flask.palletsprojects.com/en/1.1.x/blueprints/

**Comment by Eomm** [comment]
Do you have already choose a nodejs template engine? You could check the supported list by the [point-of-view](https://github.com/fastify/point-of-view#point-of-view) plugin.
EJS or TWIG should support a sort of template inheritance 

The Blueprints feature seems the same Fastify encapsulated context - for what I read right now, I did not use flask so far.
For example, you could build a plugin that exposes many routes (like `/admin` in the example you linked)

then it would be quite the same:

```
fastify.register(require('./my-admin-plugin'), { prefix: '/customerABar, applicationConfig: optsOne } )`
fastify.register(require('./my-admin-plugin'), { prefix: '/customerFoo', applicationConfig: optsTwo } )`
```

You may go deeper reading the plugins guides:

https://github.com/fastify/fastify/blob/master/docs/Plugins-Guide.md
https://github.com/fastify/fastify/blob/master/docs/Plugins.md

-- 

Auto promotion 😁: for the tutorials, I'm (trying to) writing a series to touch many aspects of fastify because I saw many people follow very old tutorials that don't follow the Fastify [documented guidelines](https://github.com/fastify/fastify/blob/master/docs/Testing.md).  
In the last one I'm writing right now I will [go deeper the encapsulation context](https://github.com/Eomm/fastify-discord-bot-demo/blob/to-production/posts/03-to-production.md#encapsulation) that would be helpful to understand if it fits the Blueprint one.

Any topic suggestions are appreciated, so you could drop a request in that repo 👍
